sdks/python/apache_beam/io/gcp/tests/bigquery_matcher.py

```diff
@@ -0,0 +1,110 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Bigquery data verifier for end-to-end test."""
+
+import logging
+
+from hamcrest.core.base_matcher import BaseMatcher
+
+from apache_beam.tests.test_utils import compute_hash
+from apache_beam.utils import retry
+
+# Protect against environments where bigquery library is not available.
+# pylint: disable=wrong-import-order, wrong-import-position
+try:
+  from google.cloud import bigquery
+  from google.cloud.exceptions import GoogleCloudError
+except ImportError:
+  bigquery = None
+# pylint: enable=wrong-import-order, wrong-import-position
+
+MAX_RETRIES = 4
+
+
+def retry_on_http_and_value_error(exception):
+  """Filter allowing retries on Bigquery errors and value error."""
+  if isinstance(exception, GoogleCloudError) or \
+          isinstance(exception, ValueError):
+    return True
+  else:
+    return False
+
+
+class BigqueryMatcher(BaseMatcher):
+  """Matcher that verifies Bigquery data with given query.
+
+  Fetch Bigquery data with given query, compute a hash string and compare
+  with expected checksum.
+  """
+
+  def __init__(self, project, query, checksum):
+    if bigquery is None:
+      raise ImportError(
+          'Bigquery dependencies are not installed.')
+    if not query or not isinstance(query, str):
+      raise ValueError(
+          'Invalid argument: query. Please use non-empty string')
+    if not checksum or not isinstance(checksum, str):
+      raise ValueError(
+          'Invalid argument: checksum. Please use non-empty string')
+    self.project = project
+    self.query = query
+    self.expected_checksum = checksum
+
+  def _matches(self, _):
+    logging.info('Start verify Bigquery data.')
+    # Run query
+    bigquery_client = bigquery.Client(project=self.project)
+    response = self._query_with_retry(bigquery_client)
+
+    # Compute checksum
+    self.checksum = compute_hash(response)
+    logging.info('Read from given query (%s), total rows %d, checksum %s',
+                 self.query, len(response), self.checksum)
+
+    # Verify result
+    return self.checksum == self.expected_checksum
+
+  @retry.with_exponential_backoff(
+      num_retries=MAX_RETRIES,
+      retry_filter=retry_on_http_and_value_error)
+  def _query_with_retry(self, bigquery_client):
+    """Run Bigquery query with retry if got error http response"""
+    query = bigquery_client.run_sync_query(self.query)
+    query.run()
+
+    # Fetch query data one page at a time.
+    page_token = None
+    results = []
+    while True:
+      rows, _, page_token = query.fetch_data(page_token=page_token)
+      results.extend(rows)
+      if not page_token:
+        break
+
+    return results
+
+  def describe_to(self, description):
+    description \
```

Comment:
This is the function inherited from hamcrest BaseMatcher. It has no return. Expected value and more related info can be added to `description` and will show in std_out when the assertion is failed.